File: Python3/_127.py
class Solution:
    def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
        wordList = set(wordList)
        if endWord not in wordList:
            return 0
        
        wordLen = len(beginWord)
        lookSet = {beginWord}
        checkSet = {endWord}
        dist = 2

        while lookSet:
            newLookSet = set()

            for word in lookSet:
                for i in range(wordLen):
                    for c in string.ascii_lowercase:

                        if c != word[i]:
                            newWord = word[:i ] + c + word[i + 1:]

                            if newWord in checkSet:
                                return dist

                            if newWord in wordList:
                                newLookSet.add(newWord)
                                wordList.remove(newWord)

            dist += 1
            lookSet = newLookSet

            if len(checkSet) < len(lookSet):
                lookSet, checkSet = checkSet, lookSet

        return 0

File: Java/_127.java
class Solution {
    public int ladderLength(String beginWord, String endWord, List<String> wordList) {

        Set<String> set = new HashSet<>(wordList);
        if (!set.contains(endWord)) return 0;

        Set<String> lookSet = new HashSet<>();
        Set<String> checkSet = new HashSet<>();

        lookSet.add(beginWord);
        checkSet.add(endWord);

        int dist = 2;

        while (!lookSet.isEmpty()) {
            Set<String> newLookSet = new HashSet<>();

            for (String str : lookSet) {
                char[] word = str.toCharArray();

                for (int i = 0; i < word.length; i++) {
                    char prevC = word[i];

                    for (char c = 'a'; c <= 'z'; c++) {
                        if (c == str.charAt(i)) continue;
                        word[i] = c;

                        String newWord = new String(word);
                        if (checkSet.contains(newWord)) return dist;

                        if (set.contains(newWord)) {
                            newLookSet.add(newWord);
                            set.remove(newWord);
                        }
                    }
                    word[i] = prevC;
                }
            }
            dist++;
            lookSet = newLookSet;

            if (checkSet.size() < lookSet.size()) {
                Set<String> temp = lookSet;
                lookSet = checkSet;
                checkSet = temp;
            }

        }
        return 0;
    }
}